#Data cleaning-Messy Employee Dataset (from kaggle)

## Introduction

*   In this notebook, I will be looking at "Messy Employee Dataset" to clean and prepare the data for future exploration and visualization

*   The goal of this project is to successfully and accurately clean the dataset, handle missing or inconsistent data, and develop a clear understanding of the data before performing deeper analysis.
*   This is an essential step in data analysis because cleaning messy data will help form insightful data and demonstrates my ability to work with real world data.

### Importing Library for data analysis

In [2]:
import pandas as pd #This is a python library that is used for data analysis

##Importing Dataset

In [5]:
df= pd.read_csv("Messy_Employee_dataset.csv") #This code shows the information of the dataset

##Looking at the information on the dataset

In [7]:
df.info() #This code gives more information on the dataset

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1020 entries, 0 to 1019
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Employee_ID        1020 non-null   object 
 1   First_Name         1020 non-null   object 
 2   Last_Name          1020 non-null   object 
 3   Age                809 non-null    float64
 4   Department_Region  1020 non-null   object 
 5   Status             1020 non-null   object 
 6   Join_Date          1020 non-null   object 
 7   Salary             996 non-null    float64
 8   Email              1020 non-null   object 
 9   Phone              1020 non-null   int64  
 10  Performance_Score  1020 non-null   object 
 11  Remote_Work        1020 non-null   bool   
dtypes: bool(1), float64(2), int64(1), object(8)
memory usage: 88.8+ KB


In this dataset, there are 10 columns and 1020 rows, with a mixture of data types including object, float, integer, and boolean. Analyzing the dataset, there are two columns with missing data, which will be addressed during the data cleaning process.

In [14]:
df.describe() #This code shows the summary statistic of each numeric column

Age         Salary         Phone
count  809.000000     996.000000  1.020000e+03
mean    32.484549   85155.056396 -4.942253e+09
std      5.656860   19873.727918  2.817326e+09
min     25.000000   50047.320000 -9.994973e+09
25%     25.000000   68392.487500 -7.341992e+09
50%     30.000000   85547.870000 -4.943997e+09
75%     40.000000  100974.027500 -2.520391e+09
max     40.000000  119971.650000 -3.896086e+06

Looking at the summary statistic, the count shows missing values for the age and salary columns, while phone shows complete rows. Age ranges from 25 to 40 with a median age of 30, showing small variability due to data showing a younger workforce. Salary ranges from $50,047-$119,971 with a median salary of $85,547, exhibiting a wider variability due to different job roles having different salaries. Looking at the standard deviation, age has moderate variation, while salary has higher variation reflecting the difference between the two.

## Looking for any duplicate values

In [23]:
df.duplicated() #This code check to see if there are any duplicate values using true or false
df[df.duplicated()] #This code out it into a dataframe and returns reselt if true

Empty DataFrame
Columns: [Employee_ID, First_Name, Last_Name, Age, Department_Region, Status, Join_Date, Salary, Email, Phone, Performance_Score, Remote_Work]
Index: []

There are no duplicated value in this data set

## Looking for null values

In [30]:
df.isnull().any() #This code check to see if there are any null values using true or false

Employee_ID          False
First_Name           False
Last_Name            False
Age                   True
Department_Region    False
Status               False
Join_Date            False
Salary                True
Email                False
Phone                False
Performance_Score    False
Remote_Work          False
dtype: bool

In [31]:
df.isnull().sum()

Employee_ID            0
First_Name             0
Last_Name              0
Age                  211
Department_Region      0
Status                 0
Join_Date              0
Salary                24
Email                  0
Phone                  0
Performance_Score      0
Remote_Work            0
dtype: int64

Looking at the dataframe, Age and Salary are the only columns with duplicate values. Age has 211 null values, and salary has 24 null values.

## Imputing missing rows for Salary and Age

In [39]:
df['Salary']=pd.to_numeric(df['Salary'],errors='coerce') #This code coverts the string into a numeric column
df['Salary'].fillna(df['Salary'].median(), inplace=True) #this code looks at the column salary and fills the missing value with the median variable)

/tmp/ipython-input-2683939953.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['Salary'].fillna(df['Salary'].median(), inplace=True) #this code looks at the column salary and fills the missing value with the median variable)


In [40]:
df['Age'].fillna(df['Age'].median(), inplace=True) #this code looks at the column age and fills the missing value with the median variable)

/tmp/ipython-input-3692653150.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['Age'].fillna(df['Age'].median(), inplace=True) #this code looks at the column age and fills the missing value with the median varibale)


In [41]:
df.info() #I am using this code to check to make sure all variables are correctly filled

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1020 entries, 0 to 1019
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Employee_ID        1020 non-null   object 
 1   First_Name         1020 non-null   object 
 2   Last_Name          1020 non-null   object 
 3   Age                1020 non-null   float64
 4   Department_Region  1020 non-null   object 
 5   Status             1020 non-null   object 
 6   Join_Date          1020 non-null   object 
 7   Salary             1020 non-null   float64
 8   Email              1020 non-null   object 
 9   Phone              1020 non-null   int64  
 10  Performance_Score  1020 non-null   object 
 11  Remote_Work        1020 non-null   bool   
dtypes: bool(1), float64(2), int64(1), object(8)
memory usage: 88.8+ KB


While imputing the missing variables, I decided to use median to fill Age and Salary columns due to both variables showing variability and the median is less sensitive to outliers

During the imputation process for salary, a data type issue occurred. The Salary column was stored as an object instead of a numeric type. To fix this issue, I used errors='coerce' to convert the column into a numeric format. Then I was able to assign NaN to invalid or missing entries in this dataset

## Changing Phone Data Type

In [42]:
df['Phone']=df['Phone'].astype(str) #This code changes the Phone column's data type from numeric to string

In [43]:
df.info() #I am using this code to check to make sure all variables are correctly filled

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1020 entries, 0 to 1019
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Employee_ID        1020 non-null   object 
 1   First_Name         1020 non-null   object 
 2   Last_Name          1020 non-null   object 
 3   Age                1020 non-null   float64
 4   Department_Region  1020 non-null   object 
 5   Status             1020 non-null   object 
 6   Join_Date          1020 non-null   object 
 7   Salary             1020 non-null   float64
 8   Email              1020 non-null   object 
 9   Phone              1020 non-null   object 
 10  Performance_Score  1020 non-null   object 
 11  Remote_Work        1020 non-null   bool   
dtypes: bool(1), float64(2), object(9)
memory usage: 88.8+ KB


For calculation purpose in the future, I changed the Phone column data type from float to string so there weren't any calculation errors to arise.

## Fixing Date and Time Column

In [44]:
df['Join_Date']=pd.to_datetime(df['Join_Date']) #This code converts the original date to python friendly time and date

In [45]:
df.info() #I am using this code to check to make sure all variables are correctly filled

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1020 entries, 0 to 1019
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dtype         
---  ------             --------------  -----         
 0   Employee_ID        1020 non-null   object        
 1   First_Name         1020 non-null   object        
 2   Last_Name          1020 non-null   object        
 3   Age                1020 non-null   float64       
 4   Department_Region  1020 non-null   object        
 5   Status             1020 non-null   object        
 6   Join_Date          1020 non-null   datetime64[ns]
 7   Salary             1020 non-null   float64       
 8   Email              1020 non-null   object        
 9   Phone              1020 non-null   object        
 10  Performance_Score  1020 non-null   object        
 11  Remote_Work        1020 non-null   bool          
dtypes: bool(1), datetime64[ns](1), float64(2), object(8)
memory usage: 88.8+ KB


I was able to successfully change the data type from a string to date and type format for future analysis

#Summary of Data Cleaning

*In this notebook, after handling missing values, correcting data types, and verifying consistency, I was able to successfully clean this dataset and prepare it for advance analysis. The Age and Salary columns were imputed using median, The Join_Date coulm was sucessfully changed into a datetime format, and Phone was changed to string to avoid any calculation errors. All duplicates and missing values have been addressed, ensurinf the dataset is complete for future exploration and modeling.

## Snapshot

In [48]:
df.info()
df.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1020 entries, 0 to 1019
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dtype         
---  ------             --------------  -----         
 0   Employee_ID        1020 non-null   object        
 1   First_Name         1020 non-null   object        
 2   Last_Name          1020 non-null   object        
 3   Age                1020 non-null   float64       
 4   Department_Region  1020 non-null   object        
 5   Status             1020 non-null   object        
 6   Join_Date          1020 non-null   datetime64[ns]
 7   Salary             1020 non-null   float64       
 8   Email              1020 non-null   object        
 9   Phone              1020 non-null   object        
 10  Performance_Score  1020 non-null   object        
 11  Remote_Work        1020 non-null   bool          
dtypes: bool(1), datetime64[ns](1), float64(2), object(8)
memory usage: 88.8+ KB


Employee_ID First_Name Last_Name   Age   Department_Region    Status  \
0     EMP1000        Bob     Davis  25.0   DevOps-California    Active   
1     EMP1001        Bob     Brown  30.0       Finance-Texas    Active   
2     EMP1002      Alice     Jones  30.0        Admin-Nevada   Pending   
3     EMP1003        Eva     Davis  25.0        Admin-Nevada  Inactive   
4     EMP1004      Frank  Williams  25.0  Cloud Tech-Florida    Active   

   Join_Date     Salary                       Email        Phone  \
0 2021-04-02   59767.65       bob.davis@example.com  -1651623197   
1 2020-07-10   65304.66       bob.brown@example.com  -1898471390   
2 2023-12-07   88145.90     alice.jones@example.com  -5596363211   
3 2021-11-27   69450.99       eva.davis@example.com  -3476490784   
4 2022-01-05  109324.61  frank.williams@example.com  -1586734256   

  Performance_Score  Remote_Work  
0           Average         True  
1         Excellent         True  
2              Good         True  
3              Good         True  
4              Poor        False